src/server/routes/api/thumbnail.js

```diff
@@ -0,0 +1,155 @@
+import Express from 'express';
+import Cytoscape from 'cytoscape';
+import VizMapper from '../../../model/vizmapper';
+import { DEFAULT_NODE_STYLE, DEFAULT_EDGE_STYLE } from '../../../model/style';
+import fetch from 'node-fetch';
+import { getNetworkDocURL, getSnapshotDocURL, removeQuotes } from './history';
+
+// TODO Rewrite this to use 'import'
+const cytosnap = require('cytosnap');
+
+console.log("Cytosnap stared!");
+
+function objMap(obj, f) {
+  return Object.fromEntries(Object.entries(obj).map(([k,v]) => [k, f(v)]));
+}
+
+async function getThumbnail(id, snapID, width, height) {
+  // If its a thumbnail of the live network document then just create and return it.
+  if(!snapID) {
+    console.log("no snapID, generating thumbnail for network doc");
+    const docURL = getNetworkDocURL(id);
+    const { image } = await createThumbnail(docURL, width, height);
+    return image;
+  }
+
+  // If its a thumbnail of a snapshot, then we can cache it as an attachment on the snapshot document.
+  const docURL = getSnapshotDocURL(id, snapID);
+
+  // Check if there's already a thumbnail attached to the snapshot document.
+  // TODO What if the width/height is different?????
```

Comment:
Not sure what the width/height comment refers to

Reply:
When a thumbnail is generated for a snapshot it is cached as an attachment in couchDB. If another request for a thumbnail for the same snapshot is sent, but for a different height/width to the cached thumbnail, then the height/width is ignored and the cached thumbnail is just returned. 

This won't be an issue for now because the History Panel always requests the same height/width. Christian needs thumbnails for the recent networks list, but those will be thumbnails for the actual network and not snapshots, so its not a problem there either. I thought it was worth noting if this needs to be fixed in the future.

I don't see any way to save extra metadata along with the attachment, which is kind of surprising to me. Maybe I just couldn't find this in the docs. Because I would like to save the height/width along with the image blob, that way we could know when to regenerate the thumbnail, or cache multiple thumbnails at different sizes.

Reply:
That's a good point.  Good to know for future

If we need different dimensions for other cases, one option is simple software scaling/cropping with CSS (e.g. background image cover / contain / etc).  That's not ideal for minimising bandwidth usage, but if the thumbnails are relatively small to begin with, then it may not be a big deal.

Another option would be to store the images as base64 text and then we can store metadata alongside it in the json doc.

A lot of trade-offs